File: kimssumin/Q31.py

def solution(numbers, target):
    answer = dfs(numbers, 0, 0, target)
    return answer

def dfs(nums, cnt, now, t):
    ans = 0
    if cnt == len(nums):
        if now == t: 
            return 1
        else: 
            return 0
    ans += dfs(nums, cnt+1, now + nums[cnt], t)
    ans += dfs(nums, cnt+1, now - nums[cnt], t)
    return ans

File: BaeYoungSuk/Q31.java
import java.util.*;
class Solution {
    
    static int count;
    static int[] copy;
    static int targetCopy;
    
    public int solution(int[] numbers, int target) {
        copy = Arrays.copyOf(numbers, numbers.length);
        count = 0;
        targetCopy = target;
        dfs(0, 0);
 
        return count;
    }
    
    public static void dfs(int depth, int sum){
        if(depth == copy.length){
            if(sum == targetCopy){
                count++;
            }
            return;
        }
        int temp = sum;

        dfs(depth+1, temp - copy[depth]);
        sum += copy[depth];
        dfs(depth+1, temp + copy[depth]);
    }
}
